crates/rooch-executor/src/actor/executor.rs

```diff
@@ -115,41 +121,58 @@ impl ExecutorActor {
-        let auth_validator_option = tx_validate_result.auth_validator();
-        match auth_validator_option {
-            Some(auth_validator) => {
-                let auth_validator_caller = AuthValidatorCaller::new(&self.moveos, auth_validator);
-                auth_validator_caller.validate(ctx, authenticator.authenticator.payload)?;
-                // pre_execute_function: TransactionValidator first, then AuthValidator
-                let pre_execute_functions = vec![
-                    TransactionValidator::pre_execute_function_call(),
-                    auth_validator_caller.pre_execute_function_call(),
-                ];
-                // post_execute_function: AuthValidator first, then TransactionValidator
-                let post_execute_functions = vec![
-                    auth_validator_caller.post_execute_function_call(),
-                    TransactionValidator::post_execute_function_call(),
-                ];
-                Ok((
-                    tx_validate_result,
-                    pre_execute_functions,
-                    post_execute_functions,
-                ))
-            }
-            None => {
-                let pre_execute_functions = vec![TransactionValidator::pre_execute_function_call()];
-                let post_execute_functions =
-                    vec![TransactionValidator::post_execute_function_call()];
-                Ok((
-                    tx_validate_result,
-                    pre_execute_functions,
-                    post_execute_functions,
-                ))
+        let tx_validate_function_result = tx_validator
+            .validate(ctx, authenticator.clone())?
+            .into_result();
+        let vm_result = match tx_validate_function_result {
+            Ok(tx_validate_result) => {
+                let auth_validator_option = tx_validate_result.auth_validator();
+                match auth_validator_option {
+                    Some(auth_validator) => {
+                        let auth_validator_caller =
+                            AuthValidatorCaller::new(&self.moveos, auth_validator);
+                        let auth_validator_function_result = auth_validator_caller
+                            .validate(ctx, authenticator.authenticator.payload)?
+                            .into_result();
+                        match auth_validator_function_result {
+                            Ok(_) => {
+                                // pre_execute_function: TransactionValidator first, then AuthValidator
+                                let pre_execute_functions = vec![
+                                    TransactionValidator::pre_execute_function_call(),
+                                    auth_validator_caller.pre_execute_function_call(),
+                                ];
+                                // post_execute_function: AuthValidator first, then TransactionValidator
```

Comment:
Why pre_execute_function: TransactionValidator first, then AuthValidator , and post_execute_function the order is reversed?

Reply:
Yes, like AOP, TransactionValidator is on the outside.